### PR TITLE
ExprPrefixSuffix Cleanup

### PR DESCRIPTION
**Pull request by Asleeepp**
### Description

I've cleaned up some bits in ExprPrefixSuffix, by adding a CompleteableFuture.runAsync, similar to ExprGroup, to prevent any possible stack traces from LuckPerms, i've also added a reset/delete changer. 

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

**Review by sovdeeth — CHANGES_REQUESTED** [review]
you should fix the method ordering while you're here, as well as switch to Jetbrains @Nullable, remove finals from method params, and rename variables to reasonable names,

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 1]
```diff
@@ -56,44 +60,57 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
-	}
-	
+
 	@Override
 	@Nullable
 	public Class<?>[] acceptChange(final ChangeMode mode) {
 		if (mode == ChangeMode.SET)
 			return new Class[] {String.class};
 		return null;
 	}
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:79]: Did you test this? You never updated the `acceptChange` method. The reset/delete modes aren't valid

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 2]
```diff
@@ -56,44 +60,57 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
+					case RESET:
+					case REMOVE:
+						if (prefix)
+							VaultHook.chat.setPlayerPrefix(player, null);
+						else
+							VaultHook.chat.setPlayerSuffix(player, null);
+						break;
+					default:
+						break;
+				}
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:101]: Use the newer switch syntax

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 3]
```diff
@@ -56,44 +60,57 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
-		}
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		CompletableFuture.runAsync(() -> {
+			for (Player player : getExpr().getArray(event)) {
+				switch (mode) {
+					case SET:
+						if (prefix)
+							VaultHook.chat.setPlayerPrefix(player, (String) delta[0]);
+						else
+							VaultHook.chat.setPlayerSuffix(player, (String) delta[0]);
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:90]:
```suggestion
						if (prefix) {
							VaultHook.chat.setPlayerPrefix(player, (String) delta[0]);
						} else {
							VaultHook.chat.setPlayerSuffix(player, (String) delta[0]);
						}
```

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 4]
```diff
@@ -56,44 +60,57 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
+						break;
+					case RESET:
+					case REMOVE:
+						if (prefix)
+							VaultHook.chat.setPlayerPrefix(player, null);
+						else
+							VaultHook.chat.setPlayerSuffix(player, null);
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:97]:
```suggestion
						if (prefix) {
							VaultHook.chat.setPlayerPrefix(player, null);
						} else {
							VaultHook.chat.setPlayerSuffix(player, null);
						}
```

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]
Could add some tests

**Comment by Asleeepp** [comment]
> Could add some tests

not sure how I do that for this, sorry 😬

**Comment by UnderscoreTud** [comment]
> > Could add some tests
> 
> not sure how I do that for this, sorry 😬

https://github.com/SkriptLang/Skript/blob/master/src/test/skript/README.md

**Comment by EquipableMC** [comment]
> > > Could add some tests
> > 
> > 
> > not sure how I do that for this, sorry 😬
> 
> https://github.com/SkriptLang/Skript/blob/master/src/test/skript/README.md

I think what he means is how would he do tests for this expression exactly

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 5]
```diff
@@ -56,44 +57,58 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
+	public String convert(Player player) {
+		return Utils.replaceChatStyles(prefix ? VaultHook.chat.getPlayerPrefix(player) : VaultHook.chat.getPlayerSuffix(player));
 	}
-	
+
 	@Override
-	protected String getPropertyName() {
-		return prefix ? "prefix" : "suffix";
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:72]:
```suggestion
	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
```

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 6]
```diff
@@ -56,44 +57,58 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
+		return switch (mode) {
+			case SET -> new Class[] {String.class};
+			case RESET, REMOVE -> new Class[] {null};
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:75]: this is illegal! the array cannot contain nulls

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 7]
```diff
@@ -56,44 +57,58 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
+			default -> null;
+		};
 	}
-	
+
 	@Override
-	public Class<? extends String> getReturnType() {
-		return String.class;
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:81]:
```suggestion
	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
```

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by UnderscoreTud — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 8]
```diff
@@ -56,44 +57,58 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
-	
+
 	@Override
-	protected String getPropertyName() {
-		return prefix ? "prefix" : "suffix";
+	@Nullable
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		return switch (mode) {
+			case SET -> new Class[] {String.class};
+			case RESET, REMOVE -> new Class<?>[0];
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:75]:
```suggestion
			case RESET, DELETE -> new Class<?>[0];
```

File: src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java [hunk 9]
```diff
@@ -56,44 +57,58 @@ public class ExprPrefixSuffix extends SimplePropertyExpression<Player, String> {
...
+			for (Player player : getExpr().getArray(event)) {
+				switch (mode) {
+					case SET -> {
+						if (prefix) {
+							VaultHook.chat.setPlayerPrefix(player, (String) delta[0]);
+						} else {
+							VaultHook.chat.setPlayerSuffix(player, (String) delta[0]);
+						}
+					}
+					case RESET, REMOVE -> {
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/hooks/chat/expressions/ExprPrefixSuffix.java:92]:
```suggestion
					case RESET, DELETE -> {
```

**Review by UnderscoreTud — DISMISSED** [review]

**Pull request merged**